src/qz/utils/GtkUtilities.java

```diff
@@ -0,0 +1,120 @@
+package qz.utils;
+
+import com.sun.jna.Library;
+import com.sun.jna.Native;
+import com.sun.jna.Pointer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Method;
+
+public class GtkUtilities {
+    private static final Logger log = LoggerFactory.getLogger(GtkUtilities.class);
+
+    /**
+     * Initializes Gtk2/3 and returns the desktop scaling factor, usually 1.0 or 2.0
+     */
+    public static double getScaleFactor() {
+        GTK gtkHandle = getGtkInstance();
+        if (gtkHandle != null && gtkHandle.gtk_init_check(0, null)) {
+            log.debug("Initialized Gtk");
+
+            if (gtkHandle instanceof GTK2) {
+                return getGtk2ScaleFactor((GTK2)gtkHandle);
+            } else {
+                return getGtk3ScaleFactor((GTK3)gtkHandle);
+            }
+        } else {
+            log.warn("An error occurred initializing the Gtk library");
+        }
+        return 0;
+    }
+
+    private static GTK getGtkInstance() {
+        log.debug("Finding preferred Gtk version...");
+        switch(getGtkMajorVersion()) {
+            case 2:
+                return GTK2.INSTANCE;
+            case 3:
+                return GTK3.INSTANCE;
+            default:
+                log.warn("Not a compatible Gtk version");
+        }
+        return null;
+    }
+
+    /**
+     * Get the major version of Gtk (e.g. 2, 3)
+     * UNIXToolkit is unavailable on Windows or Mac; reflection is required.
+     * @return Major version if found, zero if not.
+     */
+    private static int getGtkMajorVersion() {
+        try {
+            Class toolkitClass = Class.forName("sun.awt.UNIXToolkit");
+            Method versionMethod = toolkitClass.getDeclaredMethod("getGtkVersion");
+            Enum versionInfo = (Enum)versionMethod.invoke(toolkitClass);
+            Method numberMethod = versionInfo.getClass().getDeclaredMethod("getNumber");
+            int version = ((Integer)numberMethod.invoke(versionInfo)).intValue();
+            log.debug("Found Gtk{}", version);
+            return version;
+        } catch(Throwable t) {
+            log.warn("Could obtain GtkVersion information from UNIXToolkit: {}", t.getMessage());
+        }
+        return 0;
+    }
+
+    private static double getGtk2ScaleFactor(GTK2 gtk2) {
+        Pointer display = gtk2.gdk_display_get_default();
+        log.debug("Gtk 3.22+ detected, calling \"gdk_screen_get_resolution\"");
```

Comment:
Is "3.22" the right version for this method?

Reply:
should be 2.10